### PR TITLE
Mission town square v1 feature image attachment

### PR DESCRIPTION
**Pull request by SMLukwiya**
Added Image attachment feature
Added custom body parser logic

### CODE REVIEW

File: pages/api/apps/[...args].ts [hunk 1]
```diff
@@ -1,9 +1,31 @@
...
+
+  if (req.headers["content-type"] && req.headers["content-type"].indexOf("multipart/form-data") !== -1) {
+    const { fields, files } = await bodyParserEntity.parseRequestFormData(req);
+    req.body = fields;
+    req.files = files;
+  }
+
+  if (req.headers["content-type"] && req.headers["content-type"].indexOf("application/json") !== -1) {
+    req.body = await bodyParserEntity.parseRequestBufferData(req);
+  }
```
Comment (danieltonel) [inline pages/api/apps/[...args].ts:28]: I think it would be nicer to have this as part of the `bodyParserEntity` and then call:
```
const parsedRequest = bodyParserEntity.parse(req);
```

File: pages/api/apps/[...args].ts [hunk 2]
```diff
@@ -1,9 +1,31 @@
...
+    bodyParser: false,
+  },
+};
+
+export interface NextApiRequestWithFile extends NextApiRequest {
+  files: any;
+}
+
+export default async function handler(req: NextApiRequestWithFile, res: NextApiResponse) {
   const entity = new AppStoreApiProxyEntity();
```
Comment (danieltonel) [inline pages/api/apps/[...args].ts:17]: I'd rename this to `appStoreApiProxiEntity` now, since there's also another entity in this file with the `bodyParserEntity` now

File: pages/api/apps/[...args].ts [hunk 3]
```diff
@@ -1,9 +1,31 @@
...
-export default async function handler(req: NextApiRequest, res: NextApiResponse) {
+export const config = {
+  api: {
+    bodyParser: false,
+  },
+};
+
+export interface NextApiRequestWithFile extends NextApiRequest {
+  files: any;
+}
```
Comment (danieltonel) [inline pages/api/apps/[...args].ts:14]: image

What about these errors? Do you also see them? 🤔 if yes, they should have been pretty obvious to correct before requesting a review

File: app-store/apps/town-square/business-logic/message.entity.ts [hunk 4]
```diff
@@ -173,4 +186,8 @@ export default class MessageEntity {
     const pushNotification = new PushNotificationEntity();
     pushNotification.send(userIds, title, body);
   }
+
+  // private async saveToObjectStorage(file: File) {
+  //   // save to object database, return url
+  // }
```
Comment (danieltonel) [inline app-store/apps/town-square/business-logic/message.entity.ts:192]: it would have been slightly better if you added a comment via Github here. This way the reviewer sees that you didn't forget about it. Something like:

```
Todo: We'll still need to save this to an object database
```

**Comment by danieltonel** [comment]
Awesome, tested it locally, it works great! This will be really useful once we have it merged 🎉

I'll look into the object storage next

File: app-store/apps/mini-blog/pages/index.tsx [hunk 5]
```diff
@@ -29,8 +29,8 @@ export default function Index() {
             <div key={post.id} className="py-6 border-b-4 border-white border-dotted">
               <h2 className="h2">{post.title}</h2>
               <p className="paragraph">{buildPostExcerpt(post.content)}</p>
-              <Link href={`/apps/mini-blog/posts/${post.id}`}>
-                <a className="link">Read more</a>
+              <Link href={`/apps/mini-blog/posts/${post.id}`} className="link">
```
Comment (danieltonel) [inline app-store/apps/mini-blog/pages/index.tsx:32]: good catch 👌

File: app-store/apps/town-square/business-logic/message.entity.ts [hunk 6]
```diff
@@ -122,13 +123,19 @@ export default class MessageEntity {
...
 
   private async createReply(params: MessageRequest, userId: string) {
-    const { content, isReply = true } = params;
+    const { content, imageAttachment, isReply = true } = params;
 
+    let imageUrl = "";
     const sanitizedContent = sanitizeContent(content);
+    if (imageAttachment) {
+      // set image url to file name because formidable is currently saving to the public folder
+      imageUrl = imageAttachment.newFilename;
```
Comment (danieltonel) [inline app-store/apps/town-square/business-logic/message.entity.ts:132]: alternative way instead of having the `let` and an `if`:

```
const imageUrl = imageAttachment?.newFilename;
```

File: app-store/apps/town-square/components/NewMessageForm.tsx [hunk 7]
```diff
@@ -67,8 +86,35 @@ export default function MessageForm({ threadId }: { threadId?: string }) {
...
+              />
+              <Image alt="" src={imagePreviewUrl} layout="fill" objectFit="cover" className="rounded-lg" />
+            </div>
+          </div>
+        )}
+
         <div className="flex relative max-w-6xl m-auto px-1 lg:px-8 items-center">
+          <div className="mr-1">
+            <label htmlFor="file" className="cursor-pointer">
+              <div className="bg-slate-600 flex items-center justify-center h-10 w-10">
```
Comment (danieltonel) [inline app-store/apps/town-square/components/NewMessageForm.tsx:105]: Last week I did a change to replace our `bg-slate` with `bg-gray` when adding the new component library https://flowbite.com/  (they use `gray` for their default components).

Could you change the `bg-slate-600` to `bg-gray-600`?

File: app-store/apps/town-square/pages/api/messages/create.ts [hunk 8]
```diff
@@ -16,7 +18,18 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
   const entity = new MessageEntity();
 
   try {
-    const requestBody = messageRequestSchema.parse(req.body);
+    // formidable throws error if uploadDir doesn't exist
```
Comment (danieltonel) [inline app-store/apps/town-square/pages/api/messages/create.ts:21]: I like the comment, maybe make sense to additionally add the link to the related part of the documentation here because it's still a little bit confusing

is this recommended when using formidable with nextjs?

File: pages/api/apps/[...args].ts [hunk 9]
```diff
@@ -1,9 +1,31 @@
...
   const entity = new AppStoreApiProxyEntity();
+  const bodyParserEntity = new BodyParseEntity();
+
+  if (req.headers["content-type"] && req.headers["content-type"].indexOf("multipart/form-data") !== -1) {
+    const { fields, files } = await bodyParserEntity.parseRequestFormData(req);
+    req.body = fields;
+    req.files = files;
+  }
+
+  if (req.headers["content-type"] && req.headers["content-type"].indexOf("application/json") !== -1) {
```
Comment (danieltonel) [inline pages/api/apps/[...args].ts:26]: Is there a reason why you went for for `indexOf` vs using `includes`?

I think this version would be slightly more readable:

```javascript
const contentType = req.headers["content-type"];
if (contentType?.indexOf("multipart/form-data") !== -1) {
...
} else if (contentType?.indexOf("application/json") !== -1) {
   ...
} else {
  throw new Error("Not implemented");
}

```

File: pages/api/apps/[...args].ts (hunk 1 above)
Comment (SMLukwiya) [reply pages/api/apps/[...args].ts:28]: Moved this logic to bodyParserEntity class, exported one "parse" function out of the class.

File: pages/api/apps/[...args].ts (hunk 2 above)
Comment (SMLukwiya) [reply pages/api/apps/[...args].ts:17]: Completed.

File: pages/api/apps/[...args].ts (hunk 3 above)
Comment (SMLukwiya) [reply pages/api/apps/[...args].ts:14]: Changed any type to safer unknown type.

File: app-store/apps/town-square/business-logic/message.entity.ts (hunk 4 above)
Comment (SMLukwiya) [reply app-store/apps/town-square/business-logic/message.entity.ts:192]: Added a Todo comment on the code as well.

File: app-store/apps/town-square/business-logic/message.entity.ts (hunk 6 above)
Comment (SMLukwiya) [reply app-store/apps/town-square/business-logic/message.entity.ts:132]: Changed to optional chaining version

File: app-store/apps/town-square/components/NewMessageForm.tsx (hunk 7 above)
Comment (SMLukwiya) [reply app-store/apps/town-square/components/NewMessageForm.tsx:105]: Completed

File: app-store/apps/town-square/pages/api/messages/create.ts (hunk 8 above)
Comment (SMLukwiya) [reply app-store/apps/town-square/pages/api/messages/create.ts:21]: This is just a development step where we use the file system as a database. In production, formidable uses the temporary dir (os.tmpdir()) to store the image being uploaded. So we can eliminate the uploadDir options, and let formidable use the default.

File: pages/api/apps/[...args].ts (hunk 9 above)
Comment (SMLukwiya) [reply pages/api/apps/[...args].ts:26]: Changed this from "indexOf" to "include" for better readability. I excluded the final else "throw new Error("Not implemented") as it throws the error for all incoming requests.

File: pages/api/apps/[...args].ts [hunk 10]
```diff
@@ -1,9 +1,22 @@
...
+  },
+};
+
+export default async function appStoreApiProxyEntity(req: NextApiRequestWithFile, res: NextApiResponse) {
   const entity = new AppStoreApiProxyEntity();
+  const bodyParserEntity = new BodyParseEntity();
+
+  const parsedRequest = await bodyParserEntity.parse(req);
+
+  req = parsedRequest;
```
Comment (danieltonel) [inline pages/api/apps/[...args].ts:19]: I think we can skip this and just pass the `parsedRequest` laster:

```javascript
    const { apiHandler, id } = await entity.run(parsedRequest.query.args);
    parsedRequest.query.id = id;
    return apiHandler(parsedRequest, res);
```

File: app-store/apps/town-square/api-contracts/message.schema.ts [hunk 11]
```diff
@@ -29,6 +32,7 @@ export const messageListSchema = z.array(messageSchema);
 
 export const messageRequestSchema = z.object({
   content: z.string().min(1),
+  imageAttachment: isClientSide ? z.instanceof(File).or(z.instanceof(FileList)) : z.any(),
```
Comment (danieltonel) [inline app-store/apps/town-square/api-contracts/message.schema.ts:35]: can you create a `serverSideImageSchema` instead of using `any`?

```typescript 
const serverSideImageSchema = z.object({
  imageFile: z.object({
    filepath: z.string(),
    mimetype: z.string(),
    newFilename: z.string(),
    originalFilename: z.string(),
    size: z.number(),
    lastModifiedDate: z.date(),
  }),
});

export type ServerSideImage = z.TypeOf<typeof serverSideImageSchema>;
```

on the serverside you can then do:
```typescript
const imageAttachment = params.imageAttachment as ServerSideImage | undefined;
```

It's not optimal like this but I think it's better like having `any`

File: app-store/apps/town-square/business-logic/message.entity.ts (hunk 4 above)
Comment (danieltonel) [reply app-store/apps/town-square/business-logic/message.entity.ts:187]: For the image upload it would be cool to have a `image-storage.entity.ts`. And then we can just call:

```typescript
const imageStorageEntity = new ImageStorageEntity();
imageStorageEntity.save(file);
```

If it's production it should be saved to google cloud storage. If it's dev or test it should be saved to the local file system.

Comment (danieltonel) [reply app-store/apps/town-square/business-logic/message.entity.ts:187]: can you look into the upload to gcp next? you'll need to install this package `@google-cloud/storage`

here are files that you can use for reference:
- https://github.com/leerob/nextjs-gcp-storage/blob/main/pages/api/upload-url.js
- https://github.com/leerob/nextjs-gcp-storage/blob/main/.env.example

File: app-store/apps/town-square/components/NewMessageForm.tsx [hunk 12]
```diff
@@ -67,8 +86,35 @@ export default function MessageForm({ threadId }: { threadId?: string }) {
 
   return (
     <form onSubmit={form.handleSubmit(onSubmitHandler)}>
-      <div className="fixed bottom-11 left-0 w-full bg-gray-900 py-2">
+      <div className="fixed bottom-11 left-0 w-full bg-slate-900 py-2">
```
Comment (danieltonel) [inline app-store/apps/town-square/components/NewMessageForm.tsx:89]: can you change this to `bg-gray-900`?

File: app-store/apps/town-square/business-logic/message.entity.ts [hunk 13]
```diff
@@ -121,14 +127,17 @@ export default class MessageEntity {
     return await prisma.message_TownSquare.delete({ where: { id } });
   }
 
-  private async createReply(params: MessageRequest, userId: string) {
+  private createReply = async (params: MessageRequest, userId: string) => {
```
Comment (danieltonel) [inline app-store/apps/town-square/business-logic/message.entity.ts:130]: why did you change this?

File: app-store/apps/town-square/business-logic/message.entity.ts [hunk 14]
```diff
@@ -144,16 +153,23 @@ export default class MessageEntity {
     });
 
     return response;
-  }
+  };
 
-  private async createMessage(params: MessageRequest, userId: string) {
+  /**
+   * This function loses its reference to 'this' from create().
+   * Using arrow function so that reference to 'this' is not lost from line 57
```
Comment (danieltonel) [inline app-store/apps/town-square/business-logic/message.entity.ts:160]: > 'this' is not lost from line 57
once someone changes anything before line 57 this comment will be wrong

File: app-store/apps/town-square/business-logic/message.entity.ts [hunk 15]
```diff
@@ -121,14 +127,17 @@ export default class MessageEntity {
     return await prisma.message_TownSquare.delete({ where: { id } });
   }
 
-  private async createReply(params: MessageRequest, userId: string) {
+  private createReply = async (params: MessageRequest, userId: string) => {
     const { content, isReply = true } = params;
 
     const sanitizedContent = sanitizeContent(content);
+    const imageAttachment = params.imageAttachment as ServerSideImage | undefined;
+    const imageUrl = await this.saveToObjectStorage(imageAttachment);
```
Comment (danieltonel) [inline app-store/apps/town-square/business-logic/message.entity.ts:135]: do we always save to google cloud storage now? also in dev mode?

I think we shouldn't require every dev to setup their own google cloud storage bucket and just save it on the local file system instead then

Comment (danieltonel) [reply app-store/apps/town-square/business-logic/message.entity.ts:135]: we have a constant called `IS_PRODUCTION` that can be used

File: app-store/apps/town-square/business-logic/message.entity.ts (hunk 14 above)
Comment (danieltonel) [reply app-store/apps/town-square/business-logic/message.entity.ts:160]: I think it's fine to remove the comment

File: app-store/apps/town-square/business-logic/message.entity.ts (hunk 15 above)
Comment (danieltonel) [reply app-store/apps/town-square/business-logic/message.entity.ts:135]: I think this logic can be part of the imageStorageEntity then

File: business-logic/image-storage.entity.ts [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+    const fileToUpload = file.imageFile;
+
+    if (fileToUpload) {
+      const options = {
+        destination: `messages/${fileToUpload.newFilename}`,
+      };
+
+      try {
+        await bucket.upload(fileToUpload.filepath, options);
+        const imageUrl = `https://storage.googleapis.com/${bucketName}/messages/${fileToUpload.newFilename}`;
```
Comment (danieltonel) [inline business-logic/image-storage.entity.ts:26]: is this the correct and save way to get the image url?

Comment (SMLukwiya) [reply business-logic/image-storage.entity.ts:26]: Yes, but the bucket must have storageViewer access to allUsers. So that the image urls are public

File: app-store/apps/town-square/business-logic/message.entity.ts (hunk 15 above)
Comment (SMLukwiya) [reply app-store/apps/town-square/business-logic/message.entity.ts:135]: In dev, images are now saved inside the public directory

File: pages/api/apps/[...args].ts [hunk 17]
```diff
@@ -1,14 +1,25 @@
...
+import { NextApiResponse } from "next";
 
-export default async function handler(req: NextApiRequest, res: NextApiResponse) {
+export const config = {
+  api: {
+    bodyParser: false,
+  },
+};
+
+export default async function appStoreApiProxyEntity(req: NextApiRequestWithFile, res: NextApiResponse) {
```
Comment (danieltonel) [inline pages/api/apps/[...args].ts:13]: this should be called `handler`

File: app-store/apps/town-square/business-logic/message.entity.ts [hunk 18]
```diff
@@ -173,4 +185,13 @@ export default class MessageEntity {
     const pushNotification = new PushNotificationEntity();
     pushNotification.send(userIds, title, body);
   }
+
+  private async saveToObjectStorage(file?: ServerSideImage) {
```
Comment (danieltonel) [inline app-store/apps/town-square/business-logic/message.entity.ts:189]: I don't like that this is just a small helper function but has lots of complexity in it. the param can be `file` or `undefined`. The return type is `string` or `void`.

I'd only call the function if we have a file. Then we can change the param to required + the return type to always string.

```typescript
const imageFile = file?.imageFile;
const imageUrl =imageFile ? await this.saveToObjectStorage(imageFile) : undefined;
```

As a next step, I'd skip having the helper function, and change it to something like this:

```typescript
const imageFile = file?.imageFile;
const imageUrl = imageFile ? await new ImageStorageEntity().save(file) : undefined;
```

File: .gitignore [hunk 19]
```diff
@@ -42,3 +42,6 @@ npm-debug.log*
 **/public/sw.js.map
 **/public/workbox-*.js.map
 **/public/worker-*.js.map
+
+# development
+/public/devModeImages
```
Comment (danieltonel) [inline .gitignore:47]: smart 👌

File: business-logic/image-storage.entity.ts [hunk 20]
```diff
@@ -0,0 +1,39 @@
+import { ServerSideImage } from "@app-store/apps/town-square/api-contracts/message.schema";
+import { IS_PRODUCTION } from "@app-store/shared/utils/config/constants";
+import { Storage } from "@google-cloud/storage";
+
+const bucketName = process.env.GOOGLE_STORAGE_BUCKET_NAME || "";
```
Comment (danieltonel) [inline business-logic/image-storage.entity.ts:5]: let's make this uppercase `BUCKET_NAME`

File: business-logic/image-storage.entity.ts (hunk 16 above)
Comment (danieltonel) [reply business-logic/image-storage.entity.ts]: what's the default `newFilename`? If it isn't a random string already, can we make it a random string then, similar to what google drive does when you create a sharable link?

Comment (SMLukwiya) [reply business-logic/image-storage.entity.ts]: newFilename is already a random string from formidable, this is the format; 11dc85880457bdf4d81677d00.png

File: pages/api/apps/[...args].ts (hunk 17 above)
Comment (SMLukwiya) [reply pages/api/apps/[...args].ts:13]: Changed to handler

File: business-logic/image-storage.entity.ts (hunk 20 above)
Comment (SMLukwiya) [reply business-logic/image-storage.entity.ts:5]: Changed

File: business-logic/image-storage.entity.ts [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+        try {
+          await bucket.upload(fileToUpload.filepath, options);
+          const imageUrl = `https://storage.googleapis.com/${bucketName}/messages/${fileToUpload.newFilename}`;
+          return imageUrl;
+        } catch (err) {
+          console.log(err);
+        }
+      } else {
+        // image already saved locally to formidable's uploadDir (public/devModeImages)
+        return `/devModeImages/${fileToUpload?.newFilename}`;
```
[the remark this answers is not in the document]
Comment (SMLukwiya) [reply business-logic/image-storage.entity.ts:35]: Changed
